book-content/chapters/07-mutability.md

```diff
@@ -0,0 +1,1013 @@
+
+#### Solution 2: Inferring Literal Values In Arrays
+
+Let's look at some different options for solving this challenge.
+
+##### Option 1: Annotate the Entire Array
+
+The `as const` assertion can be used to solve this problem. By annotating the entire array with `as const`, TypeScript will infer the literal type of the `type` property:
+
+```typescript
+const buttonsToChange = [
+  {
+    type: "button",
+  },
+  {
+    type: "submit",
+  },
+] as const;
+```
+
+Hovering over `buttonsToChange` shows that TypeScript has inferred the `type` property as a literal type, and `modifyButtons` will no longer show an error when `buttonsToChange` is passed to it:
+
+```typescript
+// hovering over buttonsToChange shows:
+const buttonsToChange: readonly [
+  {
+    readonly type: "button";
+  },
+  {
+    readonly type: "submit";
+  },
+];
+```
+
+##### Option 2: Annotate the members of the array
+
+Another way to solve this problem is to annotate each member of the array with `as const`:
+
+```typescript
+const buttonsToChange = [
+  {
+    type: "button",
+  } as const,
+  {
+    type: "submit",
+  } as const,
+];
+```
+
+Hovering over `buttonsToChange` shows something interesting. Each object is now inferred as `readonly`, but the array itself is not:
+
+```typescript
+// hovering over buttonsToChange shows:
+const buttonsToChange: (
+  | {
+      readonly type: "button";
+    }
+  | {
+      readonly type: "submit";
+    }
+)[];
+```
+
+The `buttonsToChange` array is also no longer being inferred as a tuple with a fixed length, so we can modify it by pushing new objects to it:
+
+```typescript
+buttonsToChange.push({
+  type: "button",
+});
+```
+
+This behavior stems from tagging the individual members of the array with `as const`, instead of the entire array.
+
+However, this inference is good enough to satisfy `modifyButtons`, because it matches the `ButtonAttributes` type.
+
+##### Option 3: `as const` on strings
+
+The last solution we'll look at is using `as const` on the string literals to infer the literal type:
+
+```typescript
+const buttonsToChange = [
+  {
+    type: "button" as const,
+  },
+  {
+    type: "submit" as const,
+  },
+];
+```
+
+Now when we hover over `buttonsToChange` we've lost the `readonly` modifier, because `as const` is only being targeted at the string inside of the object, not the object itself:
+
+```typescript
+// hovering over buttonsToChange shows:
+const buttonsToChange: (
+  | {
+      type: "button";
+    }
+  | {
+      type: "submit";
+    }
+)[];
+```
+
+But again, this is still typed strongly enough to to satisfy `modifyButtons`.
+
+When using `as const` like this acts like a hint to TypeScript that it should infer a literal type where it wouldn't otherwise. This can be occasionally useful for when you want to allow mutation, but still want to infer a literal type.
```

Comment:
A neat trick is to also know how to infer a tuple using `as const` without making the tuple itself readonly. That can be done using `as const satisfies unknown[]` but you haven't introduced `satisfies` yet... so probably not worth mentioning it here